kevlar/include/kevlar_bits/grid/grid_range.hpp

```diff
@@ -28,6 +27,7 @@ struct GridRange {
     std::vector<tile_t>
         tiles_;  // vector of tiles (flattened across all gridpoints)
 
+   private:
```

Comment:
I moved this since GridRange is mostly immutable now it seemed fairly harmless to let users just grab the things they need directly. Totally understand if you want to revert. btw, why have the underscore suffix and also private storage? the underscore suffix is common python notation for a private variable, but it's not necessary when the variable is compiler protected.

Reply:
It's very common practice to denote private variables with underscore in C++ for extra clarity. Here's what I'd recommend:

- Put back `private`
- Delete the non-const getter functions. Now that I added the constructor where you can pass in `thetas, radii, sim_sizes`, there's no need to even expose the non-const version in the python exports. I think you're gonna have to change some of the C++ tests tho. 
- Replace these getter functions that are exported as `.def_readonly("thetas", &GridRange::thetas)`. Should be more Pythonic now so you can just do `grid_range.thetas` to get the underlying thetas.

In general, I never expose members as public and always expose getter functions. You have much better control of access this way. For example, if members are mutable through some of the interface, but you never want to expose these members as mutable to the outside, you can just make the getter as `const`. If you expose the member directly, there's nothing stopping the user from modifying it. You could put `const` on the member itself, but now the member fns that could modify it has to do a `const_cast` and it just logically doesn't make sense. You lose nothing by adding this extra layer of getters.

Reply:
> It's very common practice to denote private variables with underscore in C++ for extra clarity. 

Yeah, that's one of those "there are as many ways to write C++ as there are C++ developers." Happy to fit in here. I don't like the style mainly because it's noisy and distracting, but I'm not working on the C++ very much so I don't really care. 

> In general, I never expose members as public and always expose getter functions. You have much better control of access this way. 

> You lose nothing by adding this extra layer of getters.

This is rambly because I'm trying to do too much shit today... Sorry. 

I fully understand this and I've written a lot of code like this. I just don't think it's a good tradeoff and I totally disagree with the part about losing nothing. It adds extra noise to the code and starts us down the path of OOP-ify something that is much more towards the plain old data side of the "object" vs "data" spectrum. I absolutely will put mutable state as private, but I also try to avoid mutable state and write mostly [plain old data](https://en.wikipedia.org/wiki/Passive_data_structure). As an example, I'll write a lot of structs that just contain const data and are created by a builder function that is outside the class. 

One of the fundamental questions with OOP is whether you are working with an ["object" or "data"](https://stackoverflow.com/questions/23406307/whats-the-difference-between-objects-and-data-structures). (I lean towards "objects" being mis-named and a better name being something more like "actor". this would lead towards more sensible OOP where people focus on actions instead of data as the main aspect of a class) Anyway, the simplest model of "data" is a fully public struct with a bunch of members and no functions. This is beautiful to me!  Combine this with [preferring free functions over member functions](https://stackoverflow.com/questions/21028773/free-function-versus-member-function) and fun ensues. 

Again, as many ways to write C++ as there are C++ devs. I'm not going to make a thing out of this, but I'm also not going to agree with you. =)

> If you expose the member directly, there's nothing stopping the user from modifying it.

I think this is a key point. I don't worry about this *at all* and it has never come back to bite me yet. If a user wants to modify the internals of a const data structure by doing a const_cast, then they are either in dire straits and **should** be allowed to modify it or they are doing something dumb and I can't remove all their tools for suicide. This just seems like a very thin argument for adding a whole bunch of noise and distraction to the code. I'll even go a step further and argue that most C++ classes should be public by default and the whole culture of private variables is a major mistake by a significant chunk of the C++ community (again, except in the case of mutable state, which is also something to be avoided whenever possible). 

> Replace these getter functions that are exported as .def_readonly("thetas", &GridRange::thetas). Should be more Pythonic now so you can just do grid_range.thetas to get the underlying thetas.

Yeah, sounds good.

Reply:
something that's also affecting my thinking here a lot is that this is a small very early stage project. we're likely going to change a lot of stuff and we don't need a huge number of constraints. I would feel pretty differently about a mature 100k line project.

Reply:
deleted my comment above. it was too obnoxious.

Reply:
I re-privatized everything in GridRange and added a const getter for the cum_n_tiles member.